tests/eas/cpuhp.py

```diff
@@ -0,0 +1,217 @@
+        logged in a ftrace marker file for post processing. The return value is
+        the path (on the host) where the script was saved.
+
+        Example of generated script:
+        > #!/system/bin/sh
+        > echo $$ > /data/local/tmp/devlib-target/hotplug.pid
+        > while true; do
+        >     echo 0 > /sys/devices/system/cpu/cpu1/online
+        >     echo "Plug off CPU 1" > /sys/kernel/debug/tracing/trace_marker
+        >     sleep 0.4245
+        >     echo 1 > /sys/devices/system/cpu/cpu1/online
+        >     echo "Plug in CPU 1" > /sys/kernel/debug/tracing/trace_marker
+        >     sleep 0.178
+        > done
+        '''
+        script_path = '/tmp/random_cpuhp.sh'
+        shift = '    '
+        marker = cls.te.ftrace.marker_file
+        sleep_min = cls.hp_stress['sleep']['min_ms']
+        sleep_max = cls.hp_stress['sleep']['max_ms']
+        with open(script_path, 'w') as script:
+            if cls.target.os == 'android':
+                script.write('#!/system/bin/sh\n')
+            else:
+                script.write('#!/bin/sh\n')
+            cls.pid_file = cls.target.working_directory + '/hotplug.pid'
+            script.write('echo $$ > {}\n'.format(cls.pid_file))
+            script.write('while true; do\n')
+            for cpu, plug_way in cls._random_cpuhp_seq():
+                # Write in sysfs entry
+                sysfs_entry = hp._cpu_path(cls.target, cpu)
+                cmd = 'echo {} > {}\n'.format(plug_way, sysfs_entry)
+                script.write(shift + cmd)
+                # Log in trace file
+                msg = 'Plug {} CPU {}'.format('in' if plug_way else 'off', cpu)
+                script.write(shift + 'echo "{}" > {}\n'.format(msg, marker))
+                # Sleep if necessary
+                if sleep_max > 0:
+                    sleep_dur_sec = random.randint(sleep_min, sleep_max)/1000.0
+                    script.write(shift + 'sleep {}\n'.format(sleep_dur_sec))
+            script.write('done\n') # end of while
+        return script_path
+
+    @classmethod
+    def _random_cpuhp_seq(cls):
+        '''
+        Yield a consitent random sequence of CPU hotplug operations
+        
+        "Consistent" means that a CPU will be plugged-in only if it was 
+        plugged-off before (and vice versa). Moreover the state of the CPUs
+        once the sequence has completed should the same as it was before. The
+        length of the sequence (that is the number of plug operations) is 
+        defined in the hp_stress member of the class. The actual length of the
+        sequence might differ from the requested one by 1 because it's easier 
+        to implement and it shouldn't be an issue for most test cases.
+        '''
+        cur_on_cpus = cls.hotpluggable_cpus[:]
+        cur_off_cpus = []
+        max_cpus_off = cls.hp_stress['max_cpus_off']
+        i = 0
+        while i < cls.hp_stress['sequence_len'] - len(cur_off_cpus):
```

Comment:
Actually we could remove `- len(cur_off_cpus)` and use a `for` loop instead. It will work just fine but it means that the number of hotplugs in the generated sequence might be less close to `hp_stress['sequence_len']` than it is with the current implementation. It's probably not an issue though.